### ISSUE TITLE
use --help or -h to show comment info

### ISSUE THREAD

**Issue by xidianwlc** [google/python-fire#26]
Is there any way that can use "-h" or "--help" to show comment of a function or class.

example:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
:authors: wanglichao
:version: 1.0 of 2017-03-20

"""

import fire


def hello(name):
    """this is a hello test1
    """
    return 'Hello {name}!'.format(name=name)


def hello2(name):
    """this is a hello test2
    """
    return 'Hello {name}!'.format(name=name)


if __name__ == '__main__':
    fire.Fire()
```

expect result:
```
python hello.py -- --help

Usage: hello.py [OPTIONS] COMMAND [ARGS]...
             authors: wanglichao
             version: 1.0 of 2017-03-20

Options:
             -- --help
             -- --interactive
             -- --trace
             -- --verbose

Command:
             fire           inner command
             hello          this is a hello test1
             hello2         this is a hello test2
```

**Comment by dbieber** [comment]
Yes, some sort of improvements to the usage strings are in order. The other thing we may want to show is the usage for the subcommands. We'll have to prepare a full design to make sure we handle all the cases well before we start making these changes.

Currently if you run "python hello.py hello2 -- --help" then you will see the docstrings that you're asking about. But as you point out, you don't see these docstrings in the top level help.

**Comment by ilcn** [comment]
@dbieber This is something I really would love to have.  Is there active work on this? If not can I do a PR?

**Comment by dbieber** [comment]
There was someone working on usage string improvements but I believe that work has stalled; a PR would be much appreciated! Please post here what you intend to do before submitting the PR though, to give an opportunity for some design discussion.

Some things to consider: some people have long function names or long docstrings, should we truncate the docstrings?
When should the module-level docstring be shown? If Fire() is called with no args, then showing the module level docstring is probably appropriate. If Fire(fn) or Fire(cls) is called, then probably it's apppropriate to only the fn or class docstring.

**Comment by ilcn** [comment]
I think it is ok not to truncate strings -- we may miss important information if the author chooses to put some rather important stuff at the end. There can also be an option fore Fire() for max_length. 
What I would like is just parse the :param docstrings and put then after each input param shown in the usage section. This conforms with most CLI programs.  
And that's why I want to start with Fire(fn) and Fire(cls) first, and if things work then expand.

**Comment by dbieber** [comment]
1. Taking the unexpected / noisy information (type, file, line, extra docstrings) and only showing them when in verbose mode is a good change.
2. Adding an option for max_length at this time would not be appropriate; there is currently nothing to configure when using Fire, and 'max_length' isn't important enough to justify adding the first configuration. We will want to add configurations later; we should start a list of potentially configurable things before adding any such configurations though. We can create a separate issue for brainstorming this list.
3. Currently we don't parse docstrings at all. If we do start parsing docstrings, we'll need to choose a set of standards that we support. Is there a good list of standards? At google we do:

```
Args:
  name: description of argument
Returns:
  Description of return values.
```

Since there are so many docstring standards, I'm hesitant to enter the business of promoting one over the other. I much prefer the approach of presenting the docstring to the user wholesale.
That said, we can definitely improve the UI, such as by removing the label "Docstring:".

[Maybe there is a way to cover a lot of docstring formats at once, e.g. by looking for lines for which the first token is an argument name and using the remainder of that line as the argument description. I don't know how robust something like this would be.]

**Comment by dbieber** [comment]
Taking a step back, why am I hesitant to require one docstring format?
The reason is that one of the most important features of Python Fire is that it works with /any/ Python program. My favorite example of this is that you can write this simple program:

```python
from PIL import Image
import fire
fire.Fire(Image)
```

and the result is a tool similar to ImageMagick.

I don't want to make design decisions that restrict the set of Python programs with which Fire works well. This is the driving principle behind my hesitation regarding parsing docstrings.

**Comment by ilcn** [comment]
About multiple docstring formats, there are 3 major ones. rSt format is the most popular, your example is Google format, and there is also a Numpy format. Sphinx has a [docstring parser](http://www.sphinx-doc.org/en/stable/ext/autodoc.html) that I can use here. With an extension I can parse all three. 

You have a very good point. In fact there is no need to require docstring format. If the parser fails, we can just use the original docstring as a fallback. Either way you can just PIL as ImageMagick, although I don't see the reason why you would do that :D

**Comment by jtratner** [comment]
side note - Sphinx has an autodoc module that handles parsing restructured text *and* numpy format *and* google format, so if at some point fire really wanted to parse things, you could likely reuse that

**Comment by dbieber** [comment]
> If the parser fails, we can just use the original docstring as a fallback.
I like that.

**Comment by ilcn** [comment]
So if things sound good I will go ahead and do it? Sorry if I'm a bit late. Got caught up with stuff.

**Comment by dbieber** [comment]
Go for it!
Perhaps we can make the docstring parser optional in the same way that ipython is optional. If it's not available, we fall back on showing the full docstring.
Then individual tools can decide whether or not to make it a required dependency for that specific tool, but it doesn't become required for light weight use of Fire.

**Comment by dbieber** [comment]
I assume there was no progress here, but if this is something that interests you feel free to check back in, as it still would be a useful improvement!

**Comment by xeniorac** [comment]
Can I take this issue up? I am new to opensource so want to start somewhere?

**Comment by dbieber** [comment]
The help screens are in a bit of flux right now, but one thing that we'll need to do is learn how to use Sphinx's autodoc effectively.

If you're interested in writing a function that uses autodoc to take as input a function and produces as output a dictionary of information e.g. {'one-line-description': value, 'description': value, 'args': values, 'returns': value, 'raises': value} then that would be a useful way to contribute.

for reference: http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html

**Comment by xeniorac** [comment]
Thanks on it!

**Comment by nraw** [comment]
are there any updates on what's happening with this?
I wanted to bring forward fire as the next big thing in python to my colleagues but was backlashed when they saw that --help doesn't really help them as they thought it would.

**Comment by dbieber** [comment]
The fix for --help working without needing the extra -- is already checked in and will be part of the next release. I'm not able to give you a reliable estimate atm for when that release will be, unfortunately.

**Comment by safijari** [comment]
Literally just started using Fire and the additional `--` was the first stumbling block :sweat_smile:. Glad the fix is in the works.

**Comment by dbieber** [comment]
The release is happening on Friday (in two days.) Both --help and -h work in the new release, with no need for the awkward additional `--`. Apologies for the delay on getting this release out.

**Comment by dbieber** [comment]
The release is out! You can now use `--help` and `-h` normally, without the awkward extra `--`.

**Comment by alexdelorenzo** [comment]
> Some things to consider: some people have long function names or long docstrings, should we truncate the docstrings?
> When should the module-level docstring be shown? If Fire() is called with no args, then showing the module level docstring is probably appropriate. If Fire(fn) or Fire(cls) is called, then probably it's apppropriate to only the fn or class docstring.

Does Fire support showing module-level docstrings in help text now, @dbieber?

Like if you have:
```python3
"""
Description
"""

from fire import Fire


def example1() -> str:
  return 'Example1'

def example2() -> str:
  return 'Example2'


if __name__ == "__main__":
  Fire()
```

When you run `python3 example.py --help`, will the docstring text "Description" be shown in the help text?
